app/src/main/java/org/thoughtcrime/securesms/crypto/ClassicDecryptingPartInputStream.java

```diff
@@ -0,0 +1,170 @@
+ * Copyright (C) 2011 Whisper Systems
+ * 
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.thoughtcrime.securesms.crypto;
+
+import androidx.annotation.NonNull;
+
+import org.signal.core.util.logging.Log;
+import org.thoughtcrime.securesms.util.LimitedInputStream;
+import org.thoughtcrime.securesms.util.Util;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+import javax.crypto.Cipher;
+import javax.crypto.CipherInputStream;
+import javax.crypto.Mac;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.spec.IvParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+
+public class ClassicDecryptingPartInputStream {
+
+  private static final String TAG = Log.tag(ClassicDecryptingPartInputStream.class);
+
+  private static final int IV_LENGTH  = 16;
+  private static final int MAC_LENGTH = 20;
+
+  public static InputStream createFor(@NonNull AttachmentSecret attachmentSecret, @NonNull File file)
+      throws IOException
+  {
+    try {
+      if (file.length() <= IV_LENGTH + MAC_LENGTH) {
+        throw new IOException("File too short");
+      }
+
+      verifyMac(attachmentSecret, file);
+
+      FileInputStream fileStream = new FileInputStream(file);
+      byte[]          ivBytes    = new byte[IV_LENGTH];
+      readFully(fileStream, ivBytes);
+
+      Cipher          cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
+      IvParameterSpec iv     = new IvParameterSpec(ivBytes);
+      cipher.init(Cipher.DECRYPT_MODE, new SecretKeySpec(attachmentSecret.getClassicCipherKey(), "AES"), iv);
+
+      return new CipherInputStreamWrapper(new LimitedInputStream(fileStream, file.length() - MAC_LENGTH - IV_LENGTH), cipher);
+    } catch (NoSuchAlgorithmException | NoSuchPaddingException | InvalidKeyException | InvalidAlgorithmParameterException e) {
+      throw new AssertionError(e);
+    }
+  }
+
+  private static void verifyMac(AttachmentSecret attachmentSecret, File file) throws IOException {
+    Mac             mac        = initializeMac(new SecretKeySpec(attachmentSecret.getClassicMacKey(), "HmacSHA1"));
+    FileInputStream macStream  = new FileInputStream(file);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *macStream*. The resource is referenced at lines 78, 82. The resource is closed at line 97. There are other execution paths that do not contain closure statements, for example, when an if condition at line 78 evaluates to true. Ensure *macStream* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).